crates/aiken-lang/src/tipo/infer.rs

```diff
@@ -464,6 +468,142 @@ fn infer_definition(
             }))
         }
 
+        Definition::Benchmark(f) => {
+            let (typed_via, annotation) = match f.arguments.first() {
+                Some(arg) => {
+                    if f.arguments.len() > 1 {
+                        return Err(Error::IncorrectTestArity {
+                            count: f.arguments.len(),
+                            location: f
+                                .arguments
+                                .get(1)
+                                .expect("arguments.len() > 1")
+                                .arg
+                                .location,
+                        });
+                    }
+
+                    let typed_via = ExprTyper::new(environment, tracing).infer(arg.via.clone())?;
+
+                    let hydrator: &mut Hydrator = hydrators.get_mut(&f.name).unwrap();
+
+                    let provided_inner_type = arg
+                        .arg
+                        .annotation
+                        .as_ref()
+                        .map(|ann| hydrator.type_from_annotation(ann, environment))
+                        .transpose()?;
+
+                    let (inferred_annotation, inferred_inner_type) = match infer_sampler(
+                        environment,
+                        provided_inner_type.clone(),
+                        &typed_via.tipo(),
+                        &arg.via.location(),
+                    ) {
+                        Ok(result) => Ok(result),
+                        Err(err) => Err(err),
+                    }?;
+
+                    // Ensure that the annotation, if any, matches the type inferred from the
+                    // Fuzzer.
+                    if let Some(provided_inner_type) = provided_inner_type {
+                        if !arg
+                            .arg
+                            .annotation
+                            .as_ref()
+                            .unwrap()
+                            .is_logically_equal(&inferred_annotation)
+                        {
+                            return Err(Error::CouldNotUnify {
+                                location: arg.arg.location,
+                                expected: inferred_inner_type.clone(),
+                                given: provided_inner_type.clone(),
+                                situation: Some(UnifyErrorSituation::FuzzerAnnotationMismatch),
```

Comment:
Probably want to either:

- Rework the annotation / situation here so that the error message look proper
- Adjust the message to be more generic and fit both Fuzzer and Sampler